bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ChangerX2Parser.java

```diff
@@ -0,0 +1,250 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.tacmi.internal.schema;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.attoparser.ParseException;
+import org.attoparser.simple.AbstractSimpleMarkupHandler;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.tacmi.internal.schema.ChangerX2Entry.OptionType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ApiPageParser} class parses the 'changerx2' page from the CMI and
+ * maps it to the results
+ *
+ * @author Christian Niessner - Initial contribution
+ */
+@NonNullByDefault
+public class ChangerX2Parser extends AbstractSimpleMarkupHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(ChangerX2Parser.class);
+
+    static enum ParserState {
+        INIT,
+        INPUT,
+        INPUT_DATA,
+        SELECT,
+        SELECT_OPTION,
+        UNKNOWN
+    }
+
+    private final String channelName;
+    private @Nullable String curOptionId;
+    private ParserState parserState = ParserState.INIT;
+    private @Nullable String address;
+    private @Nullable String addressFieldName;
+    private @Nullable String optionFieldName;
+    private @Nullable OptionType optionType;
+    private @Nullable StringBuilder curOptionValue;
+    private Map<String, @Nullable String> options;
+
+    public ChangerX2Parser(String channelName) {
+        super();
+        this.options = new LinkedHashMap<>();
+        this.channelName = channelName;
+    }
+
+    @Override
+    public void handleDocumentStart(final long startTimeNanos, final int line, final int col) throws ParseException {
+        this.parserState = ParserState.INIT;
+        this.options.clear();
+    }
+
+    @Override
+    public void handleDocumentEnd(final long endTimeNanos, final long totalTimeNanos, final int line, final int col)
+            throws ParseException {
+        if (this.parserState != ParserState.INIT) {
+            logger.debug("Parserstate == Init expected, but is {}", this.parserState);
+        }
+    }
+
+    @Override
+    @NonNullByDefault({})
```

Comment:
NonNullByDefault should normally not be set on methods.

Reply:
We had trouble finding a proper way to annotate the `Map<>` in the (inherited) method definition properly with `@Nullable` - all our testes ended with errors like _"Illegal_ redefinition of parameter attributes, inherited method from AbstractSimpleMarkupHandler declares this parameter as 'Map<String,String>' (mismatching null _constraints)"_. When you review the discussion here in the PR it was quite a topic. Key comments on this are [here](#issuecomment-672286432) and [here](#issuecomment-677918280). If you have a better idea to solve this just enlighten me ;)

Reply:
Regarding the map I suspect it might need to be some combination of Nullable placed somewhere: 
`@Nullable Map<@Nullable String, @Nullable String>`

Reply:
I tried quite every variant making sense, but I wasn't able to get it working. This is for the variant you suggested:

<img width="807" alt="map-issue" src="https://user-images.githubusercontent.com/11214652/92648689-18c42c00-f2ea-11ea-9657-6faee27861e4.png">

When compiling on the command line with `mvn` it shows the same error...